src/main/scala/TypeCheck.scala

```diff
@@ -175,7 +175,8 @@ object TypeChecker {
     case EFloat(_) => TFloat() -> env
     case EInt(v, _) => TStaticInt(v) -> env
     case EBool(_) => TBool() -> env
-    case ERecLiteral(_) => throw RecLiteralNotInBinder(expr.pos)
+    case ERecLiteral(_) => throw NotInBinder(expr.pos, "Record Literal")
+    case EArrLiteral(_) => throw NotInBinder(expr.pos, "Array Literal")
```

Comment:
@sampsyo wirt your first concern: We do a simple well-formedness check here to make sure that array literals don't exist anywhere but inside Let binders. 

This is hack which underlines two issues: (1) Well-formedness checks should not occur inside the type checker, they should be separate. (2) The AST is more stratified that Syntax.scala lets on. There should be some way to making this more explicit.

Reply:
Got it; thanks. FWIW, although it's totally fine to rule them out here for now, I think there's no harm in record literals occurring anywhere (following the signal-vs-memory distinction mentioned below).